src/pykeen/nn/node_piece.py

```diff
@@ -685,6 +673,126 @@ def __call__(
         return len(anchors) + 1, torch.as_tensor(tokens, dtype=torch.long)
 
 
+def _random_sample_no_replacement(
+    pool: Mapping[int, Collection[int]],
+    num_tokens: int,
+) -> torch.LongTensor:
+    # randomly sample without replacement num_tokens relations for each entity
+    assignment = torch.full(
+        size=(len(pool), num_tokens),
+        dtype=torch.long,
+        fill_value=-1,
+    )
+    # TODO: vectorization?
+    for idx, this_pool in tqdm(pool.items(), desc="sampling", leave=False, unit_scale=True):
+        this_pool_t = torch.as_tensor(data=list(this_pool), dtype=torch.long)
+        this_pool = this_pool_t[torch.randperm(this_pool_t.shape[0])[:num_tokens]]
+        assignment[idx, : len(this_pool_t)] = this_pool
+    return assignment
+
+
+class PrecomputedTokenizerLoader:
+    """A loader for precomputed tokenization."""
+
+    @abstractmethod
+    def __call__(self, path: pathlib.Path) -> Tuple[Mapping[int, Collection[int]], int]:
+        """Load tokenization from the given path."""
+        raise NotImplementedError
+
+
+class GalkinPickleLoader(PrecomputedTokenizerLoader):
+    """A loader for pickle files provided by Galkin et al."""
+
+    def __call__(self, path: pathlib.Path) -> Tuple[Mapping[int, Collection[int]], int]:  # noqa: D102
+        with path.open(mode="rb") as pickle_file:
+            # contains: anchor_ids, entity_ids, mapping {entity_id -> {"ancs": anchors, "dists": distances}}
+            anchor_ids, mapping = pickle.load(pickle_file)[0::2]
+        logger.info(f"Loaded precomputed pools with {len(anchor_ids)} anchors, and {len(mapping)} pools.")
+        # normalize anchor_ids
+        anchor_map = {a: i for i, a in enumerate(anchor_ids) if a >= 0}
+        # map padding to padding
+        anchor_map[-1] = -1
+        # TODO: there are other padding tokens, e.g., -99
+        # TODO: keep distances?
```

Comment:
there are indeed several auxiliary tokens in those pre-computed vocabs that were envisioned to be sent to the transformer:

```
        # auxiliary tokens for the vocabulary
        self.NOTHING_TOKEN = -99  
        self.CLS_TOKEN = -1
        self.MASK_TOKEN = -10
        self.PADDING_TOKEN = -100
        self.SEP_TOKEN = -2
```

in the end the only used padding token is -99 [(source)](https://github.com/migalkin/NodePiece/blob/9adc57efe302919d017d74fc648f853308cf75fd/lp_rp/nodepiece_tokenizer.py#L184)

regarding distances: let's drop them for now

Reply:
317ef79